snapcraft/internal/lifecycle/_clean.py

```diff
@@ -0,0 +1,197 @@
+    # if not.
+    for part_name in part_names:
+        _verify_dependents_will_be_cleaned(part_name, part_names, step, config)
+
+    # Now we can actually clean.
+    for part_name in part_names:
+        _clean_part_and_all_dependents(
+            part_name, step, config, staged_state, primed_state)
+
+
+def _remove_directory_if_empty(directory):
+    if os.path.isdir(directory) and not os.listdir(directory):
+        os.rmdir(directory)
+
+
+def _cleanup_common_directories(config, project_options):
+    max_index = -1
+    for part in config.all_parts:
+        step = part.last_step()
+        if step:
+            index = common.COMMAND_ORDER.index(step)
+            if index > max_index:
+                max_index = index
+
+    with contextlib.suppress(IndexError):
+        _cleanup_common_directories_for_step(
+            common.COMMAND_ORDER[max_index+1], project_options)
+
+
+def _cleanup_common_directories_for_step(step, project_options, parts=None):
+    if not parts:
+        parts = []
+
+    index = common.COMMAND_ORDER.index(step)
+
+    if index <= common.COMMAND_ORDER.index('prime'):
+        # Remove the priming area.
+        _cleanup_common(
+            project_options.prime_dir, 'prime', 'Cleaning up priming area',
+            parts)
+
+    if index <= common.COMMAND_ORDER.index('stage'):
+        # Remove the staging area.
+        _cleanup_common(
+            project_options.stage_dir, 'stage', 'Cleaning up staging area',
+            parts)
+
+    if index <= common.COMMAND_ORDER.index('pull'):
+        # Remove the parts directory (but leave local plugins alone).
+        _cleanup_parts_dir(
+            project_options.parts_dir, project_options.local_plugins_dir,
+            parts)
+        _cleanup_internal_snapcraft_dir()
+
+    _remove_directory_if_empty(project_options.prime_dir)
+    _remove_directory_if_empty(project_options.stage_dir)
+    _remove_directory_if_empty(project_options.parts_dir)
+
+
+def _cleanup_common(directory, step, message, parts):
+    if os.path.isdir(directory):
+        logger.info(message)
+        shutil.rmtree(directory)
+    for part in parts:
+        part.mark_cleaned(step)
+
+
+def _cleanup_parts_dir(parts_dir, local_plugins_dir, parts):
+    if os.path.exists(parts_dir):
+        logger.info('Cleaning up parts directory')
+        for subdirectory in os.listdir(parts_dir):
+            path = os.path.join(parts_dir, subdirectory)
+            if path != local_plugins_dir:
+                try:
+                    shutil.rmtree(path)
+                except NotADirectoryError:
+                    os.remove(path)
+    for part in parts:
+        part.mark_cleaned('build')
+        part.mark_cleaned('pull')
+
+
+def _cleanup_internal_snapcraft_dir():
+    if os.path.exists(constants.SNAPCRAFT_INTERNAL_DIR):
+        shutil.rmtree(constants.SNAPCRAFT_INTERNAL_DIR)
+
+
+def clean(project_options, parts, step=None):
```

Comment:
It would be nice to reorder the functions in this file to be readable from top to bottom.
But no need to do it here, I like that kind of refactor so I can take the opportunity later to do it and document something in the CODE_STYLE.md